optuna/samplers/__init__.py

```diff
@@ -56,9 +62,37 @@ def intersection_search_space(study, ordered_dict=False):
         for param_name in delete_list:
             del search_space[param_name]
 
-    search_space = search_space or {}
+        # Retrieve cache from trial_system_attrs.
+        if trial_id is None:
+            continue
+
+        json_str = trial.system_attrs.get("intersection_search_space", None)  # type: Optional[str]
+        if json_str is None:
+            continue
+        json_dict = json.loads(json_str)
+
+        delete_list = []
+        cached_search_space = {name: dict_to_distribution(dic) for name, dic in json_dict.items()}
+        for param_name in search_space:
+            if param_name not in cached_search_space:
+                delete_list.append(param_name)
+            elif cached_search_space[param_name] != search_space[param_name]:
+                delete_list.append(param_name)
+
+        for param_name in delete_list:
+            del search_space[param_name]
 
+        break
+
+    if trial_id is not None and search_space is not None:
+        json_str = json.dumps(
+            {name: distribution_to_dict(search_space[name]) for name in search_space}
+        )
+        study._storage.set_trial_system_attr(
+            trial_id, "intersection_search_space", json_str,
```

Comment:
> The computed search_space is for the entire study while the trial_id could correspond to an arbitrary trial, i.e. not necessarily the last one. This seems a bit error prone.

Sorry, you mean there is an error case of this logic, right?

> More of a question but I'm wondering if this cannot be set on a per-study level (holds the intersection search space of the last completed trial) instead of per-trial. I'm not sure it's a cleaner design but it'd put less pressure on the storage.

In my recognition, we can't create a "per-study" level search space cache because `storage.set_study_system_attr()` is not concurrent-safe. But if we provide a user option to ensure the search space is not dynamic, we can.

Reply:
> The computed search_space is for the entire study while the trial_id could correspond to an arbitrary trial, i.e. not necessarily the last one. This seems a bit error prone.

Ah, I may understand what you mean.

| trial_id | status | search_space |
| --- | --- | --- |
| 1 | COMPLETE | `{'x': UniformDistribution(-10, 10), 'y': UniformDistribution(-5, 5)}` |
| 2 | RUNNING | None |

Assuming the above trials are stored in the storage, then we create a new trial with trial_id = 3.
`intersection_search_space(study, trial_id=3)` returns `{'x': UniformDistribution(-10, 10), 'y': UniformDistribution(-5, 5)}` and save the cache of intersection_search_space`.

| trial_id | status | search_space |
| --- | --- | --- |
| 1 | COMPLETE | `{'x': UniformDistribution(-10, 10), 'y': UniformDistribution(-5, 5)}` |
| 2 | RUNNING | None |
| 3 | COMPLETE | `{'x': UniformDistribution(-10, 10), 'y': UniformDistribution(-5, 5)}` |

After that trial_id = 2 is finished with `{'x': UniformDistribution(-10, 10)}`, then it arises inconsistency.

| trial_id | status | search_space |
| --- | --- | --- |
| 1 | COMPLETE | `{'x': UniformDistribution(-10, 10), 'y': UniformDistribution(-5, 5)}` |
| 2 | COMPLETE | `{'x': UniformDistribution(-10, 10)}` |
| 3 | COMPLETE | `{'x': UniformDistribution(-10, 10), 'y': UniformDistribution(-5, 5)}` |

The result of `intersectioin_search_space` with trial_id=4 will return `{'x': UniformDistribution(-10, 10), 'y': UniformDistribution(-5, 5)}` because of the cache of `trial_id=3`.

Reply:
As you said, it might be better to skip the creating cache if there are running trials.

Reply:
If we support this case, I'll push the following diff:

```diff
diff --git a/optuna/samplers/__init__.py b/optuna/samplers/__init__.py
index bc89bf0f..36f6dac6 100644
--- a/optuna/samplers/__init__.py
+++ b/optuna/samplers/__init__.py
@@ -113,13 +113,26 @@ def intersection_search_space(study, ordered_dict=False, trial_id=None):
         break
 
     if trial_id is not None and search_space is not None:
-        # Store an intersection_search_space cache.
-        json_str = json.dumps(
-            {name: distribution_to_dict(search_space[name]) for name in search_space}
-        )
-        study._storage.set_trial_system_attr(
-            trial_id, "intersection_search_space", json_str,
-        )
+        # Build a cache if there are on running trials.
+        # See the following discussion for the reason.
+        # https://github.com/optuna/optuna/pull/885#discussion_r404506034
+        skip_to_build_cache = False
+        for trial in study.get_trials(deepcopy=False):
+            if trial._trial_id >= trial_id:
+                break
+
+            if trial.state == optuna.structs.TrialState.RUNNING:
+                skip_to_build_cache = True
+                break
+
+        if not skip_to_build_cache:
+            # Store an intersection_search_space cache.
+            json_str = json.dumps(
+                {name: distribution_to_dict(search_space[name]) for name in search_space}
+            )
+            study._storage.set_trial_system_attr(
+                trial_id, "intersection_search_space", json_str,
+            )
 
     search_space = search_space or {}
     if ordered_dict:
```

Reply:
One concern is that if the users interrupts an optimization, a trial is likely to be stuck in a a running state.